src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+
+        private struct WriteBufferWrapper : IEquatable<WriteBufferWrapper>
+        {
+            public static readonly WriteBufferWrapper EmptyContainer = new WriteBufferWrapper(null, -1, -1, true);
+
+            private byte[] _buffer;
+            private int _offset;
+            private int _count;
+            private readonly bool _readOnly;
+
+            public WriteBufferWrapper(byte[] buffer, int offset, int count) : this(buffer, offset, count, false) { }
+            private WriteBufferWrapper(byte[] buffer, int offset, int count, bool isReadOnly) : this()
+            {
+                _buffer = buffer;
+                _offset = offset;
+                _count = count;
+                _readOnly = isReadOnly;
+            }
+
+            public byte[] ByteBuffer
+            {
+                get { return _buffer; }
+                set
+                {
+                    if (_buffer == value)
+                    {
+                        return;
+                    }
+                    if (_readOnly)
+                    {
+                        throw new ArgumentException("value");
+                    }
+                    _buffer = value;
+                }
+            }
+
+            public int Offset
+            {
+                get { return _offset; }
+                set
+                {
+                    if (_offset == value)
+                    {
+                        return;
+                    }
+                    if (_readOnly)
+                    {
+                        throw new ArgumentException("value");
+                    }
+                    _offset = value;
+                }
+            }
+
+            public int Count
+            {
+                get { return _count; }
+                set
+                {
+                    if (_count == value)
+                    {
+                        return;
+                    }
+                    if (_readOnly)
+                    {
+                        throw new ArgumentException("value");
+                    }
+                    _count = value;
+                }
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (obj is WriteBufferWrapper)
+                    return this.Equals((WriteBufferWrapper)obj);
+                return false;
+            }
+
+            public bool Equals(WriteBufferWrapper other)
+            {
+                // Doing _readOnly comparison first as main use case is comparing against
+                // BufferContainer.EmptyContainer which is the only readonly instance of BufferContainer.
+                // If either instance is the read only, then this field is sufficient for comparison.
+                if (_readOnly || other._readOnly)
+                {
+                    return _readOnly == other._readOnly;
+                }
+
+                if (other._buffer == _buffer && other._offset == _offset)
```

Comment:
NIT: this check for buffer/offset/count was just slightly strange while I was hunting for the things you were comparing to make sure you covered everything :smile: